### ISSUE TITLE
Allow Defining Maximum Node Count to Disrupt within a Duration

### ISSUE THREAD

**Issue by ccortinhas-pmi** [kubernetes-sigs/karpenter#1089]
### Description

**What problem are you trying to solve?**
We need to set a schedule where disruption may occur, and set a maximum number of disruptions during that schedule. For instance, in my cluster with 15+ Karpenter managed nodes, I wish I could allow that only a single node per day may be disrupted (I know this would not be the most cost-effective/efficient solution but would be fine with that).

To my knowledge, the current disruption budget allows us to set up a schedule where zero disruptions occur during a given schedule/duration. A workaround that we did to kind of implement our view, was to set the budget as follows:

```yaml
     budgets:
    - nodes: "10%"
    - nodes: "1"
    - nodes: "0"
      schedule: "@daily"
      duration: 1439m # 23h59m
```

Effectively, per day there's only 1 minute allowed for disruptions, during that minute only one instance is usually disrupted (2 if the first one was fast enough to instantiate and join in under a minute). Of course the downside of this workaround is that that minute will always be between 23h59 and 00h00, which may not be ideal.

It would be nice to have a configuration available that would allow us to do this without the mentioned workaround. (If there already is and I missed it please forgive me).

**How important is this feature to you?**
This feature is important to avoid successive instance disruptions, we already have configured that a maximum of one instance may be disrupted during the disruption phase, however if we have 10 nodes already "expired" they will all be recycled one by one consecutively. What we would find useful here would be the ability to for example limit daily disruptions to a configurable number of nodes, ie. 1 per day.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
> This feature is important to avoid successive instance disruptions, we already have configured that a maximum of one instance may be disrupted during the disruption phase

@ccortinhas-pmi  Can you give more detail around why you are looking to only disrupt a single instance in a given day? What's the impact to your applications when you have multiple instances disruption consecutively?

**Comment by ccortinhas-pmi** [comment]
Hello @jonathan-innis , that was just an example, one in a day, one in an hour, it doesn't even have to be one, just a configurable number (or even percentage even though we would use number).

The impact of several instances disrupting consecutively is a potential degradation of some of the services. When a new instance joins the cluster, it still has to download all the containers of the pods that it is going to hold so its services will not be available immediately. While this is happening, another instance is being disrupted and will evict another set of pods (potentially to instances that will expire next) and the cycle continues.

Additionally, and this is unrelated to Karpenter I know, we must use a container registry that throttles the number of requests, we easily hit that limit when disrupting 2 to 3 big instances (each with dozens of deployments) in a row so we would like to be sure to avoid this.

Another thing that would work for us, would be if there was a setting where we could configure a minimum time between disruptions. Is there?

**Comment by jonathan-innis** [comment]
> While this is happening, another instance is being disrupted and will evict another set of pods

Do PDBs help with this at all? It sounds like you are primarily concerned with the number of instances that we deprovision in parallel and less concerned with the total number of instances that are deprovisioned within a given time window.

> we must use a container registry that throttles the number of requests, we easily hit that limit when disrupting 2 to 3 big instances (each with dozens of deployments) in a row so we would like to be sure to avoid this

This is interesting and I hadn't heard this before. Is this the Docker upstream registry or something else?

**Comment by jonathan-innis** [comment]
> While this is happening, another instance is being disrupted and will evict another set of pods (potentially to instances that will expire next) and the cycle continues

To your point about another instance getting disrupted, doesn't your budget stop this or are the images pulling in the background and the Node still Ready? Do you mean that the Node pulls images as part of its boot-up or do you mean the CRI is pulling these images as part of the pod/container start-up?

**Comment by ccortinhas-pmi** [comment]
>  do you mean the CRI is pulling these images as part of the pod/container start-up?

Yes, I mean this, as part of the applications start-up, each application/pod will download at least three container images. Nonetheless, the node after it joins the cluster also pulls some Anypoint container images that are related to the Runtime Fabric service.
